crates/sui-core/src/consensus_adapter.rs

```diff
@@ -367,23 +375,28 @@ impl ConsensusAdapter {
             "Submitting {:?} recovered pending consensus transactions to Narwhal",
             recovered.len()
         );
-        for transaction in recovered {
-            self.submit_unchecked(transaction, epoch_store);
-        }
+        self.submit_unchecked(recovered, epoch_store);
     }
 
     fn await_submit_delay(
         &self,
         committee: &Committee,
-        transaction: &ConsensusTransaction,
+        transactions: &Vec<ConsensusTransaction>,
     ) -> (impl Future<Output = ()>, usize, usize, usize) {
-        let (duration, position, positions_moved, preceding_disconnected) = match &transaction.kind
-        {
-            ConsensusTransactionKind::UserTransaction(certificate) => {
-                self.await_submit_delay_user_transaction(committee, certificate.digest())
-            }
-            _ => (Duration::ZERO, 0, 0, 0),
-        };
+        // Use the first tx's digest to compute submit delay, after sorting the transactions by digest.
+        let mut txs_sorted = transactions.clone();
+        txs_sorted.sort_by_key(|tx| match &tx.kind {
```

Comment:
we could simply pick the minimum digest here instead of sorting and taking the first: https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.min